### DISCUSSION TITLE
Feature requests

### DISCUSSION THREAD

**Discussion by philippgille** [philippgille/gokv#107] [Ideas]
This is a general thread for feature requests. Maybe some of the feature requests initially discussed here will turn into their own discussions later.

**Comment by GwynethLlewelyn** [comment]
_\*coughs\*_, well, so let me get started on the first one...

Soooo [quoting myself](https://github.com/philippgille/gokv/issues/100#issuecomment-1107516552)...

> I'd love to see the (suggested)`List()/GetAll()` method to retrieve all keys :-) I know that's a bit beyond the purpose of the simplistic (but effective!) API, and could/should be delegated to a higher level, but the truth is that I'm completely stumped at how to implement that using the _current_ API (granted, I'd need a few more things, namely, the ability to retrieve only a _subgroup_ of keys, according to some simple criteria, or not-so-simple ones such as a nice regexp, but that's ok, for my purposes, I can do that in memory). And sure, I agree with your suggestion that the 4 basic primitives would be at the core of the `SimpleStorage` object, while additional goodies would be part of an `AdvancedStorage` object. That totally makes sense to me — especially because some backends (such as Redis or in-memory storage) would possibly work with am `AdvanceStorage`, but others might only be able to work with a `SimpleStorage`. Having both options sounds like an excellent solution!

To recap: 

1. Add support for a method (`List()`/`GetAll()`) that retrieves all keys currently held in the database:
  - just retrieve them all in no particular order (i.e. whatever the 'best' transversal algorithm might be)
  - _or_ using a simple search (just all keys starting with, say, `my-domain-key`)
  - _or_ even allowing a regexp (assuming that such a functionality _can_ be implemented at all for the many possible KV stores...)
  - _or_ use XPath/CodeQL as a query language (I'd guess that would be overkill)  
  
    Obviously, such options could also be left for users to implement on top of `List()` and not be part of `gokv` at all...

2. Add support for retrieving just the first _N_ keys

   Here I'm thinking about stores holding millions of keys, where it might make more sense to retrieve just _N_ keys and some sort of pointer/marker/cursor to key _N+1_ to be retrieved on the next call

4. Leave the core functionality of the 4 primitives as a standard for a `SimpleStorage` object, and implement  `List()`/`GetAll()` and everything else as an `AdvanceStorage` object — this would allow the two to be independently developed (I'm thinking of the ability of directly providing such functionality at the layer beyond — while most KV stores have something like that, perhaps not all have such an option, which would require additional development).

**Comment by glimchb** [comment]
to avoid major break and support backward compatibility, added `WithContext` functions for get/set/delete #136 
feel free to comment and review this is a draft

**Reply by glimchb** [reply]
@philippgille what is your view on adding CTX #136 ? thanks

**Reply by philippgille** [reply]
I'm still undecided between a hard breaking change vs. adding the new methods to the interface (or a separate one) vs just adding them without changing the interface.

Currently I think the requirements for the next version are the ones mentioned in https://github.com/philippgille/gokv/discussions/159#discussioncomment-7864394, and any bigger changes like the context can come in the next version.

I'd also like to hear more opinions about the context approach from more library users. Maybe @shawncatz as you recently created the other thread?

**Reply by glimchb** [reply]
WDYT @philippgille if we keep interface unchanged until you decide, but add new function for each implementation with context ? 
will that work ? I can rebase my change or create a new one... let me know...

**Reply by shawncatz** [reply]
> I'd also like to hear more opinions about the context approach from more library users. Maybe @shawncatz as you recently created the other thread?

Hi, I guess I would lean towards having a separate `...WithContext` interface that a user to could migrate too and then adding deprecation notices to the older impl functions.

For my use case, it wouldn't be a significant lift to cut this over if the change to `WithContext` were more abrupt, but I'm guessing it might impact others more significantly.

**Reply by glimchb** [reply]
@philippgille congrats on the new release! what is the decision on `Context` ?

**Reply by philippgille** [reply]
Yesss :slightly_smiling_face: 

Let's go with implementing the context arg in _additional_ methods for now. This allows to go store by store, having each in master already, and in case we need to do a new release while not all stores are migrated yet, the  old interface and all stores would still work. _If_ all stores are updated with the context methods, then we can still choose to make a hard cut before a release. But for now it leaves us more flexibility to go with the additional methods while keeping the old interface.

:rocket:
